alg/teca_temporal_reduction.cxx

```diff
@@ -0,0 +1,2443 @@
+        op->initialize(fill_value);
+
+        // save the operator
+        this->internals->op[array] = op;
+    }
+
+    // generate one request for each time step in the interval
+    std::vector<teca_metadata> up_reqs;
+
+    std::string request_key;
+    if (md.get("index_request_key", request_key))
+    {
+        TECA_ERROR("Failed to locate the index_request_key")
+        return up_reqs;
+    }
+
+    unsigned long req_id[2];
+    if (req_in.get(request_key, req_id))
+    {
+        TECA_ERROR("Failed to get the requested index using the"
+            " index_request_key \"" << request_key << "\"")
+        return up_reqs;
+    }
+
+    int i = this->internals->indices[req_id[0]].start_index;
+    while (i <= this->internals->indices[req_id[0]].end_index)
+    {
+        teca_metadata req(req_in);
+        req.set("arrays", req_arrays);
+        req.set(request_key, {i, i});
+        up_reqs.push_back(req);
+        i += 1;
+    }
+
+    return up_reqs;
+}
+
+// --------------------------------------------------------------------------
+const_p_teca_dataset teca_temporal_reduction::execute(
+    unsigned int port,
+    const std::vector<const_p_teca_dataset> &data_in,
+    const teca_metadata &req_in,
+    int streaming)
+{
+    (void)port;
+
+    // get the requested ind
+    std::string request_key;
+    req_in.get("index_request_key", request_key);
+    unsigned long req_id[2];
+    req_in.get(request_key, req_id);
+
+    int device_id = -1;
+    req_in.get("device_id", device_id);
+
+    if (this->get_verbose())
+    {
+        std::cerr << teca_parallel_id()
+            << "teca_temporal_reduction::execute request "
+            << req_id[0] << " device " << device_id
+            << " (" << this->internals->indices[req_id[0]].start_index
+            << " - " << this->internals->indices[req_id[0]].end_index
+            << "), reducing " << data_in.size() << ", "
+            << streaming << " remain" << std::endl;
+    }
+
+    size_t n_data = data_in.size();
+
+    // We are processing data_in in reverse order
+    // because of the average operator,
+    // more precisely because of its count variable.
+    // If we process in ascending order,
+    // the count variable is erroneously initialized many times.
+    // This is due to the fact that the partial sum and the count variable
+    // from the previous execute call is located in the last position of data_in.
+    // The order matters only for average operator.
+
+    // copy the first mesh
+    const_p_teca_cartesian_mesh mesh_in
+      = std::dynamic_pointer_cast<const teca_cartesian_mesh>(data_in[n_data-1]);
+    p_teca_cartesian_mesh mesh_out = teca_cartesian_mesh::New();
+    mesh_out->shallow_copy(
+      std::const_pointer_cast<teca_cartesian_mesh>(mesh_in));
+    p_teca_array_collection arrays_out = mesh_out->get_point_arrays();
+
+    size_t n_array = this->point_arrays.size();
+
+    // accumulate incoming values
+    for (int i = n_data-2; i >= 0; --i)
```

Comment:
What happens if the number of input datasets is < 2? this is the source of the crash on `PreIndust.ar_tag.Guan_Waliser_v2` which has only 1 time step in the last interval.

We have 3 options.
1. error out when the number of inputs is less than 2.
2. handle the case where the number of inputs < 2 by passing the input through, with type conversion applied when needed. In this case set the array to all zero or the _FillValue if one is provided.
3. implement a "complete interval only" mode in the iterators, activate it for the temporal reduction app, and error out when the number of inputs is < 2.

Option 1 could be frustrating for users, you could end up processing 30 years of data then error out in the very last time step. This is an edge case so I doubt it will come up in practice very much.  Option 2 handles the issue but puts owness on the user to interpret the result. Option 3  has the advantage in that one's always sure that all the data has been weighted the same.

I'm leaning toward 2. @taobrienlbl  thoughts on this?

Reply:
I'd lean toward 2 as well.  That's the behavior I would intuitively expect.

Reply:
Good catch @burlen! I will work on this. Thanks!